### PR TITLE
Deprecate builtin first/last aggregate function and use UDAF

### LINKED ISSUE #9957: Remove builtin aggregate function `FirstValue`
### Is your feature request related to a problem or challenge?

Follow up #9874 
I did not remove the old built-in first-value aggregate function, so I plan to remove it.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Part of #8708 
Closes #9957 #10062 

## What changes are included in this PR?

Additionally, check the signature for UDAF, and extend it to support List.
Window Aggregate function also works with UDAF surprisingly without any change about the window function

### CODE REVIEW

File: datafusion/functions-aggregate/src/first_last.rs [hunk 1]
```diff
@@ -99,19 +130,37 @@ impl AggregateUDFImpl for FirstValue {
...
                     sort_exprs.push(PhysicalSortExpr {
                         expr: e,
                         options: SortOptions {
-                            descending: !sort.asc,
-                            nulls_first: sort.nulls_first,
+                            descending: !asc,
+                            nulls_first: *nulls_first,
                         },
                     });
+                } else if let Expr::Alias(alias) = expr.as_ref() {
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/first_last.rs:152]: Filed #10074 to see how to arrange the code here, replace with `create_physical_sort_expr` or others

File: datafusion/functions-aggregate/src/first_last.rs [hunk 2]
```diff
@@ -333,359 +382,120 @@ impl Accumulator for FirstValueAccumulator {
...
 
-    pub fn with_requirement_satisfied(mut self, requirement_satisfied: bool) -> Self {
-        self.requirement_satisfied = requirement_satisfied;
-        self
-    }
+        // Construct PhysicalSortExpr objects from Expr objects:
+        let mut sort_exprs = vec![];
+        for expr in acc_args.sort_exprs {
+            if let Expr::Sort(sort) = expr {
+                if let Expr::Column(col) = sort.expr.as_ref() {
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/first_last.rs:446]: LastValue has no test case about Alias, so I keep it simple first.

File: datafusion/optimizer/src/replace_distinct_aggregate.rs [hunk 3]
```diff
@@ -92,14 +90,13 @@ impl OptimizerRule for ReplaceDistinctWithAggregate {
                 let aggr_expr = select_expr
                     .iter()
                     .map(|e| {
-                        Expr::AggregateFunction(AggregateFunction::new(
-                            AggregateFunctionFunc::FirstValue,
+                        create_first_value_expr(
```
Comment (jayzhan211) [inline datafusion/optimizer/src/replace_distinct_aggregate.rs:93]: Argument support in macro is another todo.

```rust
            // "fluent expr_fn" style function
            #[doc = $DOC]
            pub fn $EXPR_FN($($arg: Expr),*) -> Expr {
                Expr::AggregateFunction(datafusion_expr::expr::AggregateFunction::new_udf(
                    $AGGREGATE_UDF_FN(),
                    vec![$($arg),*],
                    // TODO: Support arguments for `expr` API
                    false,
                    None,
                    None,
                    None,
                ))
            }
```

**Comment by alamb** [comment]
I plan to review this tomorrow (sorry I am out this week for vacation so I have more limited bandwidth)

File: datafusion/core/src/physical_optimizer/convert_first_last.rs [hunk 4]
```diff
@@ -199,15 +200,15 @@ fn try_convert_first_last_if_better(
...
+        let aggr_req = PhysicalSortRequirement::from_sort_exprs(aggr_sort_expr);
         let reverse_aggr_req =
-            PhysicalSortRequirement::from_sort_exprs(&reverse_aggr_req);
-
-        if let Some(first_value) = aggr_expr.as_any().downcast_ref::<FirstValue>() {
-            let mut first_value = first_value.clone();
+            PhysicalSortRequirement::from_sort_exprs(&reversed_ordering);
 
+        if aggr_fun_expr.is_some_and(|e| e.fun().name() == "FIRST_VALUE") {
+            let mut first_value = aggr_fun_expr.unwrap().clone();
```
Comment (jayzhan211) [inline datafusion/core/src/physical_optimizer/convert_first_last.rs:211]: where derive(Clone) is needed

File: datafusion/functions-aggregate/src/first_last.rs [hunk 5]
```diff
@@ -333,359 +363,120 @@ impl Accumulator for FirstValueAccumulator {
...
-    fn order_bys(&self) -> Option<&[PhysicalSortExpr]> {
-        (!self.ordering_req.is_empty()).then_some(&self.ordering_req)
-    }
-
-    fn name(&self) -> &str {
-        &self.name
-    }
-
-    fn reverse_expr(&self) -> Option<Arc<dyn AggregateExpr>> {
-        Some(Arc::new(self.clone().convert_to_last()))
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/first_last.rs:487]: Note that `reverse_expr` is removed and manually converted in `OptimizeAggregateOrder`, but `get_best_fitting_window` may call `reverse_expr` of first/last but no test is covered

**Comment by jayzhan211** [comment]
I think reverse_expr should be included before deprecate builtint first/last, so convert to draft.

**Pull request closed without merging**